diff --git a/lox/Parser.java b/lox/Parser.java
index bd45be7..8eb5b15 100644
--- a/lox/Parser.java
+++ b/lox/Parser.java
@@ -23,7 +23,19 @@ class Parser {
   }
 
   private Expr expression() {
-    return equality();
+    return comma();
+  }
+
+  private Expr comma() {
+    Expr expr = equality();
+
+    while (match(COMMA)) {
+      Token operator = previous();
+      Expr right = equality();
+      expr = new Expr.Binary(expr, operator, right);
+    }
+
+    return expr;
   }
 
   private Expr equality() {
